src/vs/platform/mcp/common/mcpResourceScannerService.ts

```diff
@@ -0,0 +1,124 @@
+	readonly _serviceBrand: undefined;
+	scanMcpServers(mcpResource: URI): Promise<IScannedMcpServers>;
+	addMcpServers(servers: { server: IScannedMcpServer; inputs?: IMcpServerVariable[] }[], mcpResource: URI): Promise<IScannedMcpServer[]>;
+	removeMcpServers(serverNames: string[], mcpResource: URI): Promise<void>;
+}
+
+export class McpResourceScannerService extends Disposable implements IMcpResourceScannerService {
+	readonly _serviceBrand: undefined;
+
+	private readonly resourcesAccessQueueMap = new ResourceMap<Queue<any>>();
+
+	constructor(
+		@IFileService private readonly fileService: IFileService,
+		@IUriIdentityService protected readonly uriIdentityService: IUriIdentityService,
+	) {
+		super();
+	}
+
+	async scanMcpServers(mcpResource: URI): Promise<IScannedMcpServers> {
+		return this.withProfileMcpServers(mcpResource);
+	}
+
+	async addMcpServers(servers: { server: IScannedMcpServer; inputs?: IMcpServerVariable[] }[], mcpResource: URI): Promise<IScannedMcpServer[]> {
+		const result: IScannedMcpServer[] = [];
+		await this.withProfileMcpServers(mcpResource, scannedMcpServers => {
+			let updatedInputs = scannedMcpServers.inputs ?? [];
+			const existingServers = scannedMcpServers.servers ?? {};
+			for (const { server, inputs } of servers) {
+				existingServers[server.name] = server;
+				result.push(server);
+				if (inputs) {
+					const existingInputIds = new Set(updatedInputs.map(input => input.id));
+					const newInputs = inputs.filter(input => !existingInputIds.has(input.id));
+					updatedInputs = [...updatedInputs, ...newInputs];
+				}
+			}
+			return { servers: existingServers, inputs: updatedInputs };
+		});
+		return result;
+	}
+
+	async removeMcpServers(serverNames: string[], mcpResource: URI): Promise<void> {
+		await this.withProfileMcpServers(mcpResource, scannedMcpServers => {
+			for (const serverName of serverNames) {
+				if (scannedMcpServers.servers?.[serverName]) {
+					delete scannedMcpServers.servers[serverName];
+				}
+			}
+			return scannedMcpServers;
+		});
+	}
+
+	private async withProfileMcpServers(file: URI, updateFn?: (data: IScannedMcpServers) => IScannedMcpServers): Promise<IScannedMcpServers> {
+		return this.getResourceAccessQueue(file)
+			.queue(async () => {
+				let scannedMcpServers: IScannedMcpServers | undefined;
+				try {
+					const content = await this.fileService.readFile(file);
+					const errors: ParseError[] = [];
+					scannedMcpServers = parse(content.value.toString(), errors, { allowTrailingComma: true, allowEmptyContent: true });
+					if (errors.length > 0) {
+						throw new Error('Failed to parse scanned MCP servers: ' + errors.join(', '));
+					}
+				} catch (error) {
+					if (toFileOperationResult(error) !== FileOperationResult.FILE_NOT_FOUND) {
+						throw error;
+					}
+				}
+				if (updateFn) {
+					scannedMcpServers = updateFn(scannedMcpServers ?? {});
+					if (scannedMcpServers.servers && Object.keys(scannedMcpServers.servers).length > 0) {
```

Comment:
I think we should also check for inputs here, I would be annoyed if the inputs I created were removed while I was swapping out MCP servers.